### PR TITLE
Experiments: add matcher for min/max versions

### PR DESCRIPTION
**Pull request by travis79**
- Add experiments matcher functionality for min and max versions
- Add VersionString utility class to aid in comparing versions
- Update tests to for coverage of matchers

### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### CODE REVIEW

File: components/service/experiments/README.md [hunk 1]
```diff
@@ -178,12 +178,14 @@ The experiments records in Kinto contain the following properties:
...
+| match.locale_language     |        |          | The default locales language.                                                                                                                   | "eng|zho|deu"                                                  |
+| match.app_display_version | regex  |          | The application version.                                                                                                                        | `"1.0.0"`                                                      |
+| match.app_min_version     | string |          | The minimum application version, inclusive.                                                                                                     | `"1.0.0"`                                                      |
+| match.app_max_version     | string |          | The maximum application version, inclusive.                                                                                                     | `"1.0.0"`                                                      |
```
Comment (georgf) [inline components/service/experiments/README.md]: Nit: It would be nice to use different versions for these, e.g. "1.0.1" & "1.4.2", to highlight how those could be used.

File: components/service/experiments/src/main/java/mozilla/components/service/experiments/util/VersionString.kt [hunk 2]
```diff
@@ -0,0 +1,73 @@
+package mozilla.components.service.experiments.util
+
+import java.lang.IllegalArgumentException
+import kotlin.math.max
+
+internal class VersionString(
```
Comment (georgf) [inline components/service/experiments/src/main/java/mozilla/components/service/experiments/util/VersionString.kt]: Isn't there a lot of overlap to [`EngineVersion`](https://github.com/mozilla-mobile/android-components/blob/master/components/concept/engine/src/main/java/mozilla/components/concept/engine/utils/EngineVersion.kt) here?
We shouldn't duplicate logic if we can avoid it. Why not share this as single utility code somewhere?

File: components/service/experiments/src/test/java/mozilla/components/service/experiments/ExperimentEvaluatorTest.kt [hunk 3]
```diff
@@ -251,6 +251,64 @@ class ExperimentEvaluatorTest {
...
+
+        packageInfo.versionName = "1.1.18"
+        assertNotNull(evaluator.evaluate(context, ExperimentDescriptor("testexperiment"), listOf(experiment), 20))
+
+        packageInfo.versionName = "0.1.18"
+        assertNull(evaluator.evaluate(context, ExperimentDescriptor("testexperiment"), listOf(experiment), 20))
+    }
+
+    @Test
+    fun `evaluate appMaxVersion`() {
```
Comment (georgf) [inline components/service/experiments/src/test/java/mozilla/components/service/experiments/ExperimentEvaluatorTest.kt]: Please also add a test for evaluating both min & max version together, both matching and non-matching.
Please also test on the edges of the possible value spaces, e.g. for `evaluate appMinVersion` with min=1.0.0, test for version being: 0.9.9, 1.0.0, 1.0.1, somewhere further away like 1.1.18.

File: components/service/experiments/src/main/java/mozilla/components/service/experiments/util/VersionString.kt (hunk 2 above)
Comment (travis79) [reply components/service/experiments/src/main/java/mozilla/components/service/experiments/util/VersionString.kt]: I'll file a follow-up bug and discuss it with the A-C team

Comment (travis79) [reply components/service/experiments/src/main/java/mozilla/components/service/experiments/util/VersionString.kt]: I was worried about creating a cross-dependency right now

Comment (travis79) [reply components/service/experiments/src/main/java/mozilla/components/service/experiments/util/VersionString.kt]: Filed the bug to follow-up [here](https://bugzilla.mozilla.org/show_bug.cgi?id=1562241).  I don't think that this blocks or hurts us to proceed with this functionality living in service-experiments for now.

File: components/service/experiments/src/test/java/mozilla/components/service/experiments/ExperimentEvaluatorTest.kt (hunk 3 above)
Comment (travis79) [reply components/service/experiments/src/test/java/mozilla/components/service/experiments/ExperimentEvaluatorTest.kt]: I expanded the test cases to better cover edge cases and close cases.

File: components/service/experiments/src/main/java/mozilla/components/service/experiments/util/VersionString.kt [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    override fun compareTo(other: VersionString): Int {
+        if (!version.matches(VERSION_REGEX)) {
+            throw IllegalArgumentException("Unexpected format in VersionString")
+        }
+        if (!other.version.matches(VERSION_REGEX)) {
+            throw IllegalArgumentException("Unexpected format in VersionString")
+        }
+        val thisParts = version.split(".")
+        val otherParts = other.version.split(".")
+        val length = max(thisParts.count(), otherParts.count())
```
Comment (georgf) [inline components/service/experiments/src/main/java/mozilla/components/service/experiments/util/VersionString.kt]: Your code supports versions with different element counts, but i don't see test coverage for that scenario.
Let's add that before landing.

File: components/service/experiments/src/main/java/mozilla/components/service/experiments/util/VersionString.kt [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+        for (i in 0..length) {
+            val thisPart = if (i < thisParts.count()) {
```
Comment (georgf) [inline components/service/experiments/src/main/java/mozilla/components/service/experiments/util/VersionString.kt]: Non-blocking feedback: filling the `*Parts` lists up with `0`s before the loop would probably make for more readable code.

Comment (travis79) [reply components/service/experiments/src/main/java/mozilla/components/service/experiments/util/VersionString.kt]: Added a comment for now, to better explain what's going on.

File: components/service/experiments/src/main/java/mozilla/components/service/experiments/util/VersionString.kt (hunk 4 above)
Comment (travis79) [reply components/service/experiments/src/main/java/mozilla/components/service/experiments/util/VersionString.kt]: No problem, added a handful of tests for both shorter and longer element counts
